rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -901,69 +922,125 @@ void PlayerImpl::load_storage_content()
-  // to support play_next() API logic.
-  rosbag2_storage::SerializedBagMessageSharedPtr message_ptr = peek_next_message_from_queue();
-  { // Notify play_next() that we are ready for playback
-    // Note: We should do notification that we are ready for playback after peeking pointer to
-    // the next message. message_queue_.peek() is not allowed to be called from more than one
-    // thread concurrently.
+  { // Notify play_next()/seek() that we are ready for playback
     std::lock_guard<std::mutex> lk(ready_to_play_from_queue_mutex_);
     is_ready_to_play_from_queue_ = true;
     ready_to_play_from_queue_cv_.notify_all();
   }
-  while (rclcpp::ok() && !stop_playback_ &&
-    message_ptr != nullptr && !shall_stop_at_timestamp(message_ptr->recv_timestamp))
-  {
-    // Do not move on until sleep_until returns true
-    // It will always sleep, so this is not a tight busy loop on pause
-    while (rclcpp::ok() && !clock_->sleep_until(message_ptr->recv_timestamp)) {
-      if (std::atomic_exchange(&cancel_wait_for_next_message_, false)) {
-        break;
+  rosbag2_storage::SerializedBagMessageSharedPtr message_ptr = take_next_message_from_queue();
+
+  // While we haven't stopped playing
+  while (rclcpp::ok() && !stop_playback_) {
+    // If we had run out of messages before but are starting to play next again, e.g., after a
+    // seek()
+    if (play_next_.load()) {
+      std::lock_guard<std::mutex> lk(skip_message_in_main_play_loop_mutex_);
+      skip_message_in_main_play_loop_ = false;
+      cancel_wait_for_next_message_ = false;
+      // Only take again if we need
+      if (nullptr == message_ptr) {
+        message_ptr = take_next_message_from_queue();
       }
     }
-    std::lock_guard<std::mutex> lk(skip_message_in_main_play_loop_mutex_);
-    if (rclcpp::ok()) {
-      if (skip_message_in_main_play_loop_) {
-        skip_message_in_main_play_loop_ = false;
-        cancel_wait_for_next_message_ = false;
-        message_ptr = peek_next_message_from_queue();
-        continue;
+
+    // While there's a message to play and we haven't reached the end timestamp yet
+    while (rclcpp::ok() && !stop_playback_ &&
+      message_ptr != nullptr && !shall_stop_at_timestamp(message_ptr->recv_timestamp))
+    {
+      // Sleep until the message's replay time, do not move on until sleep_until returns true
+      // It will always sleep, so this is not a tight busy loop on pause
+      // However, skip sleeping if we're trying to play the next message
+      while (rclcpp::ok() && !stop_playback_ && !play_next_.load() &&
+        !clock_->sleep_until(message_ptr->recv_timestamp))
+      {
+        // Stop sleeping if cancelled
+        if (std::atomic_exchange(&cancel_wait_for_next_message_, false)) {
+          break;
+        }
       }
-      publish_message(message_ptr);
+
+      std::lock_guard<std::mutex> lk(skip_message_in_main_play_loop_mutex_);
```

Comment:
Nitpick from clangtidy. Rename to avoid shadowed variable names in the inner scope with another `lk` definition.
```suggestion
      std::lock_guard<std::mutex> lock_skip_message(skip_message_in_main_play_loop_mutex_);
```

Reply:
renamed in ca23e7fae501e6a2950f529ef0dafa2f37dd1f26